Show not-yet-compared answers in ranked list

kitsook opened the pull request:
When viewing ranked answers:
- for instructors, all comparable answers (including those
not-yet-compared) should be included
- for students, only ranked answers within the display limit will be
shown

Fix GET `/api/courses/:course_id/assignments/:assignment_id/answers`
API:
- instructors and TAs can see ranks/scores no matter what `orderBy` is
specified
- students can only see ranks if:
  (1) assignment has `rank_display_limit` set; and,
  (2) `orderBy` equals `score`

When retrieving specific answer by ID, don't include rank/score if user
is restricted.

Closes #662, closes #657

andrew-gardener commented on compair/tests/api/test_answers.py:
```diff
@@ -121,28 +122,39 @@ def test_get_all_answers(self):
             self.assertEqual(len(expected), rv.json['total'])
 
             # test sorting by rank (display_rank_limit None)
+            # When there is no rank limit, students can't see ranks.
```
instead it should be a 400 error (as mentioned)

andrew-gardener commented on compair/tests/api/test_answers.py:
```diff
@@ -239,6 +252,22 @@ def test_get_all_answers(self):
+            self.assertEqual(2, rv.json['pages'])
+            self.assertEqual(20, rv.json['per_page'])
+            self.assertEqual(len(expected), rv.json['total'])
+
```
It would be good to fetch the second page here as well since that is where all the uncompressed answers should be

andrew-gardener commented on compair/tests/api/test_answers.py:
```diff
@@ -924,8 +953,19 @@ def test_get_user_answers(self):
             # test unsaved query
             rv = self.client.get(url, query_string={'unsaved': True})
             self.assert200(rv)
+            # The answer API compares timestamps ("modified" vs "created") to determine if an answer is "unsaved".
```
Can you actually modify the answer modal and endpoint so it uses `modified == created` and `draft == true`. This was a situation I didn't handle/catch when adding it (if a new non-draft answer is added somehow, it shouldn't be unsaved)

Thanks!

kitsook replied:
@andrew-gardener  The API supports bool parameters `unsaved` and `draft`.  If the model's `saved` property is changed to depend on `draft` property, it will be contradictory to invoke the API with `unsaved` without `draft`.

So, will change the logic to skip `draft` filter if  `unsaved` is specified.

Does it make sense?

&nbsp; | unsaved | unsaved not specified
--- | --- | ---
**draft** | not Answer.saved | Answer.draft = True
**draft not specified** | not Answer.saved | Answer.draft = False

The model will be like:
```
    @hybrid_property
    def saved(self):
        return not (self.modified == self.created and self.draft)

    @saved.expression
    def saved(cls):
        return not_(and_(cls.modified == cls.created, cls.draft))
```

And API will be like:
```
        if params.get('unsaved'):
            query = query.filter(not_(Answer.saved))
        else:
            # move the draft parameter checking here
            query = query.filter_by(draft=params.get('draft'))
```

andrew-gardener replied:
`unsaved` param is used only as workaround for getting uuids to work with xAPI (need to uuid to exist before the answer would normally). Its currently only called in one place and I want to try to find a better way of handling it in the future. 

I don't think its worth building to much around it (I think you can just use the this part):
```
        if params.get('unsaved'):
            query = query.filter(not_(Answer.saved))
```
for the endpoint

Also I prefer not using negation around multiple argurments if possible (makes it harder to follow)
```
    @hybrid_property
    def saved(self):
        return self.modified != self.created or not self.draft

    @saved.expression
    def saved(cls):
        return or_(cls.modified != cls.created, cls.draft == False)
```

andrew-gardener commented on data/fixtures/test_data.py:
```diff
@@ -556,7 +557,7 @@ def __init__(self):
 
     def add_course(self, num_students=5, num_assignments=1, num_additional_criteria=0, num_groups=0, num_answers='#',
-            with_comments=False, with_draft_student=False, with_comparisons=False, with_self_eval=False):
+            with_comments=False, with_draft_student=False, with_comparisons=False, with_self_eval=False, num_non_comparable_ans=0):
```
have `num_non_comparable_ans` set to at least 1 by default (so there is better test coverage overall)

andrew-gardener commented on data/fixtures/test_data.py:
```diff
@@ -571,7 +572,7 @@ def add_course(self, num_students=5, num_assignments=1, num_additional_criteria=
         self.assignment = self.assignments[0]
 
-        self.add_answers(num_answers, with_comments=with_comments)
+        self.add_answers(num_answers, with_comments=with_comments, num_non_comparable_ans=num_non_comparable_ans)
```
recommend instead of adding it to `add_answers`, just add the new answers in this function below (like draft student)

lenglund commented:
Would like to keep the "Instructor" label in the metadata above the answer, even for the comparable answers (unless there's a reason we would want to remove it?).

Also wondering if we can by default reset to the first page when we change the "Order by" sorting. Otherwise, the user clicks to change the sort, then has to scroll down and click first to get to the first page of answers...

kitsook commented:
The frontend has logic (`notScoredEnd`) to push those comparable answers without scores to the bottom of the page.  This gives an illusion that the answers are no longer sorted by date.  Suggest to remove such filter.

https://github.com/ubc/compair/blob/b114f39d04e3104332771e3bc84a043cab318ed8/compair/static/modules/assignment/assignment-view-partial.html#L215

kitsook commented:
>Would like to keep the "Instructor" label in the metadata above the answer, even for the comparable answers (unless there's a reason we would want to remove it?).

The "Teaching Assistant" and "Instructor" labels also serve to indicate those are non-comparable "model answers" and listed on top of the list.  If we apply "TA" and "Instructor" labels to both comparable and non-comparable answers, maybe we also need a new label for non-comparable answers?

kitsook commented:
Changes made:
- changed sorting logic in API to put non-comparable answers on top of the list when sort by date
- jump back to page 1 when "order by" filter changed
- added "Instructor Answer (not compared)" in student view
- removed the filter that pushes comparable answers without score to the bottom of the page
- included "Instructor" / "Teacher Assistant" labels to comparable answers unless anonymized the view

>Does removing the filter change anything else?

@lenglund Should be fine.  It is an AngularJS filter and not visible by users.
Student's view can't see score at all, so shouldn't be affected.  After the filter is removed, instructor's view (when ordered by date) will not be affected by whether the comparable answer has score or not.

lenglund commented:
`changed sorting logic in API to put non-comparable answers on top of the list when sort by date
Just to check here: non-comparable instructor answers are 1) at the top of the list when sorted by date and 2) removed from the list when sorted by answer score. Additionally: comparable instructor answers 3) appear according to date in the list when sorted by date and 4) appear according to score in the list when sorted by score AND compared and 5) appear at the bottom of the list when sorted by score AND not yet compared. Is this correct?

kitsook commented:
> Just to check here: non-comparable instructor answers are 1) at the top of the list when sorted by date and 2) removed from the list when sorted by answer score. Additionally: comparable instructor answers 3) appear according to date in the list when sorted by date and 4) appear according to score in the list when sorted by score AND compared and 5) appear at the bottom of the list when sorted by score AND not yet compared. Is this correct?

All correct 👍
